src/WLR/Types/Pointer.hsc

```diff
@@ -0,0 +1,135 @@
+        (#poke struct wlr_pointer, data) ptr $ wlr_pointer_data t
+
+{- 
+ - TODO there were some source comments here that had some sort of struct
+ - type specified, I guess it's a hint about the types???
+ - I think I did the comment structure so that `struct wlr_pointer_motion_event` will show on hover
+ - need to check that when I get my hls working ~ Matt t
+-}
+data WLR_pointer_events = WLR_pointer_events {
+    -- |struct wlr_pointer_motion_event
+    wlr_pointer_events_motion :: WL_signal
+    -- |struct wlr_pointer_motion_absolute_event
+    , wlr_pointer_events_motion_absolute :: WL_signal
+    -- |struct wlr_pointer_button_event
+    , wlr_pointer_events_button :: WL_signal
+    -- |struct wlr_pointer_axis_event
+    , wlr_pointer_events_axis :: WL_signal
+    , wlr_pointer_events_frame :: WL_signal
+
+    -- |struct wlr_pointer_swipe_begin_event
+    , wlr_pointer_events_swipe_begin :: WL_signal
+    -- |struct wlr_pointer_swipe_update_event
+    , wlr_pointer_events_swipe_update :: WL_signal
+    -- |struct wlr_pointer_swipe_end_event
+    , wlr_pointer_events_swipe_end :: WL_signal
+
+    -- |struct wlr_pointer_pinch_begin_event
+    , wlr_pointer_events_pinch_begin :: WL_signal
+    -- |struct wlr_pointer_pinch_update_event
+    , wlr_pointer_events_pinch_update :: WL_signal
+    -- |struct wlr_pointer_pinch_end_event
+    , wlr_pointer_events_pinch_end :: WL_signal
+
+    -- |struct wlr_pointer_hold_begin_event
+    , wlr_pointer_events_hold_begin :: WL_signal
+    -- |struct wlr_pointer_hold_end_event
+    , wlr_pointer_events_hold_end :: WL_signal
+    }
+
+instance Storable WLR_pointer_events where
+    -- Because this 'events' struct is defined within the pointer struct,
+    -- does that mean that I can't use this #alignment???
+    -- I doubt this works, but it compiles without sizeOf and alignment
+    -- maybe I have to use the argument to `alignment` TODO look into that,
+    -- maybe I can use that to tell it where to find the nested struct definition
+    --alignment _ = #alignment struct wlr_pointer.events
+    --sizeOf _ = #size struct wlr_pointer.events
+    peek ptr = WLR_pointer_events
+        <$> (#peek struct wlr_pointer, events.motion) ptr
+        <*> (#peek struct wlr_pointer, events.motion_absolute) ptr
+        <*> (#peek struct wlr_pointer, events.button) ptr
+        <*> (#peek struct wlr_pointer, events.axis) ptr
+        <*> (#peek struct wlr_pointer, events.frame) ptr
+        <*> (#peek struct wlr_pointer, events.swipe_begin) ptr
+        <*> (#peek struct wlr_pointer, events.swipe_update) ptr
+        <*> (#peek struct wlr_pointer, events.swipe_end) ptr
+        <*> (#peek struct wlr_pointer, events.pinch_begin) ptr
+        <*> (#peek struct wlr_pointer, events.pinch_update) ptr
+        <*> (#peek struct wlr_pointer, events.pinch_end) ptr
+        <*> (#peek struct wlr_pointer, events.hold_begin) ptr
+        <*> (#peek struct wlr_pointer, events.hold_end) ptr
```

Comment:
This implementation for `instance Storable WLR_pointer_events` is incorrect, to my understanding — the intention is to define it for `wlr_pointer.events`, but these `#peek`s will extract values from a `struct wlr_pointer`!

Reply:
This was so weird for me because the struct is defined in the other struct.

I thought that, since it was defined that way, it was impossible to have a wlr_pointer_events that wasn't inside a wlr_pointer, so these macros were grabbing the wlr_pointer_events from the 'parent' wlr_pointer. I thought that was what this `events.motion` dot-access syntax was doing

I tried looking at the hsroots code to figure out what to do [link](https://github.com/swaywm/hsroots/blob/eeacc3ff9cb9940e8a38c0d740a39242897ab059/src/Graphics/Wayland/WlRoots/Input/Pointer.hsc#L20)
```haskell
pointerGetEvents :: Ptr WlrPointer -> PointerEvents
pointerGetEvents ptr = PointerEvents
    { pointerButton = #{ptr struct wlr_pointer, events.button} ptr
    , pointerMotion = #{ptr struct wlr_pointer, events.motion} ptr
    , pointerMotionAbs = #{ptr struct wlr_pointer, events.motion_absolute} ptr
    , pointerAxis = #{ptr struct wlr_pointer, events.axis} ptr
    }
```

Reply:
Oh okay I think I see what you mean

Reply:
If you wish to maintain the 1:1 C struct to Haskell datatype ratio, the correct way would be to `peek` and `poke` `WLR_pointer_events` together with the rest of the data structure. So, no separate `Storable WLR_pointer_events` instance.

Reply:
Indeed, that is what I’ve been doing elsewhere in the code (e.g. [for `WLR_renderer`](https://github.com/bradrn/wlhs/blob/ffed09f8165082f4a0ecd4f0234081803ff528ec/src/WLR/Render/Renderer.hsc#L17-L18)).

Reply:
Thanks, I removed `wlr_pointer_events` and moved the peek 'n poke code into the `wlr_pointer` instance